File: OCMDAdDemo/OCMDAdDemo/DemoControllers/CustomViewController.m
//
//  CustomViewController.m
//  OCMDAdDemo
//
//  Created by DerekYuYi on 2019/10/29.
//  Copyright © 2019 RuiYu. All rights reserved.
//


/**
Abstract: 演示自定义广告类 `RYCustomAd` 的使用.
*/


#import "CustomViewController.h"
#import "CustomAdView.h"
@import MDAd;

@interface CustomViewController ()<RYCustomAdDelegate, CustomAdViewDelegate>

/// 你可以选择不同的广告类型进行自定义, 比如 Banner 广告, 信息流广告, 全屏广告等
/// 详情请看 `BannerViewController`, `InfoFlowViewController`, `InterstitialViewController` 等.
/// 这里选择信息流广告类型(850008) 来进行自定义展示.
@property (nonatomic, strong) RYCustomAd *customAd;

@property (nonatomic, strong) CustomAdView *customView;

@end

@implementation CustomViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.title = @"自定义广告";
    
    // setup your custom view
    [self setupCustomView];
    
    // setup customAd instance
    self.customAd = [[RYCustomAd alloc] initWithAdsID:@"850008"];
    self.customAd.rootViewController = self;
    self.customAd.delegate = self;
    [self.customAd loadRequest];
}

- (void)viewDidLayoutSubviews {
    [super viewDidLayoutSubviews];
    self.customView.frame = CGRectMake(0, 100, self.view.bounds.size.width, 100);
}

- (void)setupCustomView {
    
    NSArray *nibArray = [[NSBundle mainBundle] loadNibNamed:@"CustomAdView" owner:self options:nil];
    if (nibArray.firstObject && [nibArray.firstObject isKindOfClass:CustomAdView.class]) {
        self.customView = nibArray.firstObject;
        [self.view addSubview:self.customView];
        self.customView.delegate = self;
    }
}


#pragma mark - RYCustomAdDelegate

- (void)customAdDidReceiveAd:(RYCustomAd *)customAd receivedData:(RYCustomAdItem *)item {
    
    /** `140.0/180.0` 是自定义广告对应的广告图片的比例. 因为这里选择的广告是 850008, 其广告图片对应的比例是 `140.0/180.0`.
        注意: 在进行自定义广告时, 请务必保证自定义图片的比例与对应广告类型图片比例保持一致. 对应广告类型图片比例请查看对应产品文档.
    */
    [self.customView updateWithImageUrlString:item.imageUrls.firstObject
                                        title:item.title
                                         desc:item.desc
                                   imageRatio:140.0/180.0];
}

- (void)customAdDidFailToReceiveAd:(RYCustomAd *)customAd error:(RYError *)error {
    UIAlertController *alertVC = [UIAlertController alertControllerWithTitle:@"检测返回异常"
                                                                     message: error.errorDescription
                                                              preferredStyle:UIAlertControllerStyleAlert];
    UIAlertAction *action = [UIAlertAction actionWithTitle:@"确定"
                                                     style:UIAlertActionStyleDefault
                                                   handler:^(UIAlertAction * _Nonnull action) {
        [alertVC dismissViewControllerAnimated:YES completion:nil];
    }];
    [alertVC addAction:action];
    [self presentViewController:alertVC animated:YES completion:nil];
}


#pragma mark - CustomAdViewDelegate

- (void)adTapped {
    // open ad link.
    // called when you add gestures, actions or other Selectors to you custom ad view.
    [self.customAd openAdLink];
}

@end

File: SwiftMDAdDemo/SwiftMDAdDemo/Controllers/CustomViewController.swift
//
//  CustomViewController.swift
//  SwiftMDAdDemo
//
//  Created by DerekYuYi on 2019/10/18.
//  Copyright © 2019 RuiYu. All rights reserved.
//

import UIKit
import MDAd

/**
 Abstract: 演示自定义广告类 `RYCustomAd` 的使用.
 */

class CustomViewController: UIViewController {
    
    /// 你可以选择不同的广告类型进行自定义, 比如 Banner 广告, 信息流广告, 全屏广告等, 详情请看 `BannerViewController`, `InfoFlowViewController`, `InterstitialViewController` 等.
    /// 这里选择信息流广告类型(850008) 来进行自定义展示.
    fileprivate let customAd = RYCustomAd(adsID: "850008")
    
    /// 自定义广告对应的广告图片的比例. 注意: 在进行自定义广告时, 请务必保证自定义图片的比例与对应广告类型图片比例保持一致.
    fileprivate let imageViewRatioForCustomAd: CGFloat = 140.0 / 180.0
    
    /// 广告自定义视图, 你可以开发你所需要的自定义设计. 这里以 `CustomView` 为例.
    fileprivate let customView = CustomView.loadFromNib()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        title = "自定义广告示例"
        
        // add CustomView
        view.addSubview(customView)
        customView.delegate = self

        // config instance of RYCustomAd
        customAd.rootViewController = self
        customAd.delegate = self
        customAd.loadRequest()
    }
    
    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        customView.frame = CGRect(x: 0, y: 100, width: view.bounds.width, height: 100)
    }
}

// MARK: - RYCustomAdDelegate

extension CustomViewController: RYCustomAdDelegate {
    func customAdDidReceiveAd(_ customAd: RYCustomAd, receivedData infos: [String : String]) {
        
        // update datasource for customview after received custom ad data.
        customView.update(infos["MDAdKeyForImageUrl"],
                          title: infos["MDAdKeyForTitle"],
                          desc: infos["MDAdKeyForDescription"],
                          imageRatio: imageViewRatioForCustomAd)
    }
    
    func customAdDidFailToReceiveAd(_ customAd: RYCustomAd, error: RYError) {
        let alertVC = UIAlertController(title: "检测返回异常", message: error.errorDescription ?? "无详细错误信息", preferredStyle: .alert)
        let action = UIAlertAction(title: "确定", style: UIAlertAction.Style.default) { _ in
            alertVC.dismiss(animated: true, completion: nil)
        }
        alertVC.addAction(action)
        present(alertVC, animated: true, completion: nil)
    }
}

// MARK: - CustomViewDelegate

extension CustomViewController: CustomViewDelegate {
    func adTapped() {
        
        // open ad link.
        // called when you add gestures, actions or other Selectors to you custom ad view
        customAd.openAdLink()
    }
}
